### PR TITLE
[Form] Added Form view parent method

### PR DESCRIPTION
**Pull request by bhavin-nakrani**
| Q | A |
| --- | --- |
| Branch? | "master" for new features |
| Bug fix? | no |
| New feature? | yes |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #18882 |
| License | MIT |
| Doc PR | reference to the documentation PR, if any |

Adding FormView::hasParent() which should be hardcoded in twig to prevent accessing the property: form.hasParent.

### CODE REVIEW

**Comment by ogizanagi** [comment]
Just asking, but what about `isRoot()` instead, as in `FormInterface::isRoot()` ?

**Comment by bhavin-nakrani** [comment]
@ogizanagi ``` FormInterface::isRoot() ``` is not accessible from twig and ``` FormView ``` does not implement this interface.

**Comment by bhavin-nakrani** [comment]
Does anyone knows why i am getting this error in build ```AppVeyor build failed```

**Comment by HeahDude** [comment]
The problem here is that it does not make such sense to add methods like `isRoot` or `hasParent` on a value object with public properties as `FormView`, and using `form.parent` or `form.root` has the same effect in Twig that checking the public property so adding it to hardcoding it looks weird somehow.

The real issue I guess come from Twig and how it tries to access an attributes because `FormView` has both public properties and array access, but the last has a precedence so it will first look for named children.

Maybe we should just better document how to access children via offset get when they have the same name as a view property (e.g `form['parent']`, `form['children']`, `form['vars']`) and suggest to avoid such collision?

File: src/Symfony/Component/Console/Tests/Helper/ProcessHelperTest.php [hunk 1]
```diff
@@ -44,6 +44,9 @@ public function testPassedCallbackIsExecuted()
         $this->assertTrue($executed);
     }
 
+    /**
+     * @return array
+     */
     public function provideCommandsAndOutput()
```
Comment (iltar) [inline src/Symfony/Component/Console/Tests/Helper/ProcessHelperTest.php]: Your IDE can already determine the return type, in this state it doesn't really add a value

Comment (HeahDude) [inline src/Symfony/Component/Console/Tests/Helper/ProcessHelperTest.php]: This fix does not belong to the PR's scope anyway.

**Comment by ogizanagi** [comment]
> @ogizanagi FormInterface::isRoot() is not accessible from twig and FormView does not implement this interface.

@bhavin-nakrani : I was asking about naming the method `isRoot` instead of `hasParent` and reverse the logic inside 😅 

> The real issue I guess come from Twig and how it tries to access an attributes because FormView has both public properties and array access, but the last has a precedence so it will first look for named children.
>
> Maybe we should just better document how to access children via offset get when they have the same name as a view property (e.g form['parent'], form['children'], form['vars']) and suggest to avoid such collision?

@HeahDude : IIUC, this won't solve the mentioned issue, where `form.parent` will always return something in case the form has a child named `parent`. Thus, no way to know efficiently if it's a child or a parent :/

**Comment by HeahDude** [comment]
@ogizanagi Yes what I implicitly suggest here is to think about the precedence in Twig, checking object (properties and methods) in an early return before checking for array and `ArrayAccess`, but that sounds like a big BC break (maybe doable with the upcoming 2.0?).

The opposite is not working either because if one wants to test if a form has a child named "parent" it may return the actual parent even using `form['parent']` because `[]` has the same behavior for objects since it is the only way to make the call dynamic using a twig variable to determine the property/method.

So it may worth to say a word in the doc about it anyway. Although hardcoding `hasMethod` might help to ensure the checks both ways but IMO it keeps looking like a hack adding this method in `FormView` to work around this.

**Comment by ro0NL** [comment]
Imo. it's worth reserving `form.parent` as is. A child named liked like that can be accessed via `form.children['parent']`, same for having `form.root` as proposed in #20369

**Comment by fabpot** [comment]
Closing this PR as there does not seem to have any more activity and the PR cannot be merged as is. Feel free to reopen though.
